pkg/controller/controller_test.go

```diff
@@ -1090,8 +1198,8 @@ func provisionTestcases() (int64, map[string]provisioningTestcase) {
 	}
 	vacName := "test-vac"
 	return requestedBytes, map[string]provisioningTestcase{
-		"normal provision": {
-			volOpts: controller.ProvisionOptions{
+		"normal provision": *provisionTestcaseBuilder().volOpts(
```

Comment:
I think the builder pattern should be applied to things we want to build in an object to then test it externally. With this implementation both the things to build and the expectations are part of the same object.

Maybe `provisioningTestCase` shouldn't have setters for expectations.

Reply:
Thank you for your feedback. Let me confirm your suggestion.

We should create another function to set the expectations. The function is to set the following elements of the `provisioningTestcase` struct, which is the expectation of the test result.
```
	expectedPVSpec            *pvSpec
	expectedMockReturn        func(t *testing.T, out *csi.CreateVolumeResponse, controllerServer *driver.MockControllerServer, createVolumeError error)
	expectErr                 bool
	expectState               controller.ProvisioningState
	expectSelectedNode        string // a specific selected-node of the PVC in the apiserver after the test, same as before if empty
	expectNoProvision         bool   // if true, then ShouldProvision should return false
```

At the same time, the `provisionTestcase()` function should be used to build the test case, which means to set the following elements of the `provisioningTestcase` struct.
```
	capacity                  int64 // if zero, default capacity, otherwise available bytes
	volOpts                   controller.ProvisionOptions
	clientSetObjects          []runtime.Object
	createVolumeError         error
	withExtraMetadata         bool
	deploymentNode            string // fake distributed provisioning with this node as host
	immediateBinding          bool   // enable immediate binding support for distributed provisioning
	controllerPublishReadOnly bool
	featureGates              map[featuregate.Feature]bool
	pluginCapabilities        func() (rpc.PluginCapabilitySet, rpc.ControllerCapabilitySet)
```

Did I understand your suggestion properly?

Reply:
I split the first three test cases into builder (`buildProvisionTestcase()`) and expectation setter (`setProvisionTestcaseExpectation()`). Is this what you intended?

Or is the intention to further divide the `provisioningTestcase struct` into two, as shown below, and rewrite related parts based on that?
```
type provisioningTestcase struct {
	capacity                  int64 // if zero, default capacity, otherwise available bytes
	volOpts                   controller.ProvisionOptions
	clientSetObjects          []runtime.Object
	createVolumeError         error
	withExtraMetadata         bool
	deploymentNode            string // fake distributed provisioning with this node as host
	immediateBinding          bool   // enable immediate binding support for distributed provisioning
	controllerPublishReadOnly bool
	featureGates              map[featuregate.Feature]bool
	pluginCapabilities        func() (rpc.PluginCapabilitySet, rpc.ControllerCapabilitySet)
}

type provisioningTestcaseExpectation struct {
	expectedPVSpec            *pvSpec
	expectedMockReturn        func(t *testing.T, out *csi.CreateVolumeResponse, controllerServer *driver.MockControllerServer, createVolumeError error)
	expectErr                 bool
	expectState               controller.ProvisioningState
	expectSelectedNode        string // a specific selected-node of the PVC in the apiserver after the test, same as before if empty
	expectNoProvision         bool   // if true, then ShouldProvision should return false
}
```